www/js/lib/zimArchive.js

```diff
@@ -190,12 +232,15 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
         }, true).then(function(firstIndex) {
             var dirEntries = [];
             var addDirEntries = function(index) {
-                if (index >= firstIndex + resultSize || index >= that._file.articleCount)
+                if (search.state === 'cancelled' || index >= firstIndex + resultSize || index >= that._file.articleCount)
                     return dirEntries;
```

Comment:
@mossroy I think I found (one?) cause of the race condition in #184: this return statement (which was in the original code like this) should *not* return an array of dirEntries: I think it should return a resolved Promise with the array.... As can be seen from the next line, `addDirEntries` is a Promise that should only return inside the `.then` function, or by Resolving the Promise.

Reply:
OK, I'm wrong here, sorry... Have found this:

Inside a `then` handler function, when x is a value (number, string, etc):
* `return x` is equivalent to `return Promise.resolve(x)`
* `throw x` is equivalent to `return Promise.reject(x)`

https://stackoverflow.com/a/47168159/9727685